### PR TITLE
ADBDEV-5473 Truncate temporary relation files before unlink

### PR DESCRIPTION
**Pull request by dnskvlnk**
Truncate temporary relation files before unlink

Problem description:
After the drop of a temporary heap relation, disk space was not released to the
OS for several minutes. The expected behavior is immediate disk space release.

Root cause:
Background writer process kept file descriptors of the relation open until the
next checkpoint. The next checkpoint occurred after a timeout, which could be up
to CheckPointTimeout seconds (300 by default). Only temporary relations suffered
from it, because in mdunlinkfork segment files were truncated for all
non-temporary relations.

Fix:
The mdunlinkfork function was changed to call do_truncate on all segment files
for all relations, including temporary.

Modified due to conflicts:
src/test/regress/input/tablespace.source
src/test/regress/output/tablespace.source

(cherry picked from commit https://github.com/arenadata/gpdb/commit/d622b980ce2b0eb3c2fa63b8c4f1ec29d7c01930)

### CODE REVIEW

File: src/test/regress/output/tablespace.source [hunk 1]
```diff
@@ -737,3 +737,78 @@ SELECT c.relname, t.spcname FROM pg_class c LEFT JOIN pg_tablespace t ON c.relta
 
 DROP TABLE tablespace_part;
 DROP TABLESPACE myts;
+DROP ROLE tablespace_testuser1;
+DROP ROLE tablespace_testuser2;
```
Comment (silent-observer) [inline src/test/regress/output/tablespace.source]: This is not present in the input

File: src/test/regress/input/tablespace.source [hunk 2]
```diff
@@ -333,3 +333,28 @@ SELECT c.relname, t.spcname FROM pg_class c LEFT JOIN pg_tablespace t ON c.relta
...
+-- pages by the bg writer process.
+CREATE EXTENSION IF NOT EXISTS gp_inject_fault;
+-- Start with a clean state (no dirty buffers) by performing a checkpoint.
+-- It will also protect from a random timed checkpoint during the test, that could lead to not stable test result.
+CHECKPOINT;
+-- Force flushing all dirty pages by the bg writer process.
+SELECT gp_inject_fault_infinite('bg_buffer_sync_default_logic', 'skip', dbid) FROM gp_segment_configuration WHERE role = 'p' AND content > -1;
+
+-- Checking that the temp heap table doesn't leave non-zero files.
+CREATE TEMP TABLE t(a int, b int) TABLESPACE testspace;
```
Comment (silent-observer) [inline src/test/regress/input/tablespace.source]: The test fails because testspace doesn't exist

**Review by silent-observer — COMMENTED** [review]
This should be cherry-picked with -x to preserve previous commit hash in the message, or you can add it manually

File: src/test/regress/input/tablespace.source (hunk 2 above)
Comment (dnskvlnk) [reply src/test/regress/input/tablespace.source]: Fixed

File: src/test/regress/output/tablespace.source (hunk 1 above)
Comment (dnskvlnk) [reply src/test/regress/output/tablespace.source]: Fixed

**Comment by whitehawk** [comment]
The authorship of the original commit is lost.

File: src/test/regress/output/tablespace.source [hunk 3]
```diff
@@ -737,3 +737,63 @@ SELECT c.relname, t.spcname FROM pg_class c LEFT JOIN pg_tablespace t ON c.relta
...
+-- Force flushing all dirty pages by the bg writer process.
+SELECT gp_inject_fault_infinite('bg_buffer_sync_default_logic', 'skip', dbid) FROM gp_segment_configuration WHERE role = 'p' AND content > -1;
+ gp_inject_fault_infinite 
+--------------------------
+ Success:
+ Success:
+ Success:
+(3 rows)
+
+CREATE TABLESPACE testtablespace LOCATION '/home/gpadmin/gpdb_src/src/test/regress/testtablespace';
```
Comment (silent-observer) [inline src/test/regress/output/tablespace.source:756]: Please use `@testtablespace@` here too

**Pull request closed without merging**